Disallow index.time_series.end_time setting from being set or updated in normal indices

martijnvg opened the pull request:
The `index.mode` setting validates other index settings. When updating the `index.time_series.end_time` setting and the `index.mode` setting isn't wasn't defined at index creation time (meaning that default is active), then this validation is skipped which results into (worse) errors at a later point in  time.

This problem is fixed by enforced by making `index.mode` setting a dependency of `index.time_series.end_time` setting.

Note that this problem doesn't exist for the `index.time_series.start_time` and `index.routing_path` index settings, because these index settings are final, which mean these can only be defined when an index is being created. 

Closes #110265

martijnvg committed:
Trigger index.mode validation when updating index settings and the setting hasn't been defined.

The `index.mode` setting validates other index settings.
When updating the `index.routing_path` or `index.time_series.*` settings and the `index.mode` setting isn't already defined or present in put settings api, then validation is skipped which results into (worse) errors at a later point in  time.

Closes #110265

elasticsearchmachine commented:
Hi @martijnvg, I've created a changelog YAML for you.

martijnvg committed:
Update docs/changelog/110268.yaml

martijnvg committed:
fix changelog

martijnvg committed:
improve yaml test

martijnvg commented on server/src/main/java/org/elasticsearch/index/IndexSettings.java:
```diff
@@ -1157,6 +1157,14 @@ public synchronized boolean updateIndexMetadata(IndexMetadata indexMetadata) {
+        // Also in the case the index.mode hasn't been specified in newSettings (and default applies)
+        // (Note when creating a new index this happens when we read the 'mode' field in the constructor of this class
+        // regardless of whether the index.mode setting has been specified)
+        IndexSettings.MODE.get(newSettings);
```
Parsing the index.mode setting with new settings triggers the validation. We don't need to keep the result around.

I chose this place to do this because in the case when index.time_series.end_time is defined at index creation time and the index.mode setting isn't defined then the parsing of the index.mode setting in the constructor of this class triggers that validation that fails creating the index. So the logic used at index creation time and setting update time is at least in the same class.

lkts replied:
Can we do it somewhere here https://github.com/elastic/elasticsearch/blob/dea593db3f4d7d40ed23c0574a6a3f3fb8b3470e/server/src/main/java/org/elasticsearch/index/IndexSettings.java#L620?

salvatore-campagna replied:
Maybe yes if we add the `MODE` to `settings()` below as we need the `inde.mode` to be available in the settings to validate it against?

martijnvg replied:
This seems to work out. I like this approach better as the logic is contained with in the `TIME_SERIES_END_TIME` setting anonymous inner class.

elasticsearchmachine commented:
Pinging @elastic/es-storage-engine (Team:StorageEngine)

martijnvg committed:
updated changelog title

martijnvg committed:
iter

kkrik-es commented:
Some tests need fixing, they probably updated the setting in standard mode.

martijnvg committed:
adjust unit tests to deal with the fact that index.time_series.end_time is read, then index.mode index setting must be present as well.

In production this should always be the case, but for changed unit tests in this commit not.

martijnvg committed:
Merge remote-tracking branch 'es/main' into index_mode_setting_validation_bug

martijnvg committed:
spotless

martijnvg committed:
take component template settings into account when validation index template

martijnvg committed:
Merge remote-tracking branch 'es/main' into index_mode_setting_validation_bug

martijnvg committed:
removed newline

martijnvg committed:
added comments to unit test
